File: src/main/scala/com/jackywong/scala/parsing/my/Parsers.scala
package com.jackywong.scala.parsing.my

import scala.util.matching.Regex

/**
  * Created by huangziqi on 2019/7/1
  */
trait Parsers {
  def defaultSucceed[A](a: A): Parser[A]
  def succeed[A](a: A): Parser[A]
  def string(s: String): Parser[String]
  def char(c: Char): Parser[Char]
  def charIn(chars:Char*): Parser[Char]
  def charNotIn(chars:Char*):Parser[Char]
  def many1[A](p: Parser[A]): Parser[List[A]]
  def many[A](p: Parser[A]): Parser[List[A]]
  def regex(r: Regex): Parser[String]
  def whitespace: Parser[String]
  def digits: Parser[String]
  def thru(s: String): Parser[String]
  def quoted: Parser[String]
  def escaped: Parser[String]
  def escapedQuoted: Parser[String]
  def doubleString: Parser[String]
  def double: Parser[Double]
  def token[A](p: Parser[A]): Parser[A]
  def opL[A](p: Parser[A])(op: Parser[(A,A) => A]): Parser[A]
  def surround[A](start: Parser[_], stop: Parser[_])(p: => Parser[A]): Parser[A]
  def eof: Parser[String]
  def root[A](p: Parser[A]): Parser[A]
}

File: src/main/java/com/jackywong/parsing/Parsers.java
package com.jackywong.parsing;

import java.util.List;
import java.util.function.BiFunction;
import java.util.function.Supplier;

/**
 * Created by huangziqi on 2019/7/16
 */
public interface Parsers {
    <A> Parser<A> defaultSucceed(A a);
    <A> Parser<A> succeed(A a);
    Parser<String> string(String s);
    Parser<Character> pchar(Character c);
    Parser<Character> charIn(Character... c);
    Parser<Character> charNotIn(Character... c);
    <A> Parser<List<A>> many1(Parser<A> p);
    <A> Parser<? super List<A>> many(Parser<A> p);
    Parser<String> regex(String r);
    Parser<String> whitespace();
    Parser<String> digits();
    Parser<String> thru(String s);
    Parser<String> quoted();
    Parser<String> escaped();
    Parser<String> escapedQuoted();
    Parser<String> doubleString();
    Parser<Double> pdouble();
    <A> Parser<A> token(Parser<A> p);
    <A> Parser<A> opL(Parser<A> p, Parser<BiFunction<A,A,A>> op);
    <A> Parser<A> surround(Parser<?> start, Parser<?> stop, Supplier<Parser<A>> p);
    Parser<String> eof();
    <A> Parser<A> root(Parser<A> p);
}
